File: binary_search_tree.rb
require 'ruby-progressbar'

class BST
	attr_accessor :root

	def traverse(&blk)
		_traverse(@root, &blk)
	end

	def insert(node)
		if @root.nil?
			@root = node
		else
			_insert(node, @root)
		end
	end

	def find(v)
		traverse { |n| break n if n.value == v}
	end

	private

	def _insert(node, parent)
		if node.value < parent.value
			if parent.left.nil?
				parent.left = node
			else
				_insert(node, parent.left)
			end
		else
			if parent.right.nil?
				parent.right = node
			else
				_insert(node, parent.right)
			end
		end
	end

	def _traverse(node, &blk)
		return if node.nil?

		_traverse(node.left, &blk)

		yield node

		_traverse(node.right, &blk)
	end
end

class Node
	attr_accessor :value, :left, :right

	def initialize(v)
		@value = v
		@left = nil
		@right = nil
	end

	def to_s
		"v = #{@value} | l = #{@left || 'nil' } | r = #{@right || 'nil'}"
	end
end

def make_random_bst(num_nodes)
	bst = BST.new

	progress = ProgressBar.create(
		total: num_nodes
	)
	puts 'Building tree...'
	num_nodes.times { |i|
		bst.insert(Node.new(rand(i)))
		progress.increment
	}
	bst
end

# bst = make_random_bst(1000)
bst = BST.new

bst.insert(Node.new(5))
bst.insert(Node.new(2))
bst.insert(Node.new(7))
bst.insert(Node.new(9))

puts bst.root

bst.traverse { |n| puts n.value }

File: binary_search_tree.py
import random

class Node:
	def __init__(self, value):
		self.value = value
		self.left = None
		self.right = None

class BST:
	def __init__(self):
		self.root = None

	def insert(self, node):
		if self.root == None:
			self.root = node
		else:
			self.__insert(node, self.root)

	def __insert(self, node, root):
		print("inserting")
		if node.value < root.value:
			if root.left == None:
				root.left = node
			else:
				self.__insert(node, root.left)
		else:
			if root.right == None:
				root.right = node
			else:
				self.__insert(node, root.right)

	def traverse_inorder(self, fn):
		self.__traverse_inorder(self.root, fn)

	def __traverse_inorder(self, root, fn):
		if root == None:
			return

		self.__traverse_inorder(root.left, fn)
		fn(root)
		self.__traverse_inorder(root.right, fn)



def print_node(n):
	print(n.value)

bst = BST()

inputs = [random.randint(1, 100) for _ in range(0, 100)]
for i in inputs:
	bst.insert(Node(i))

bst.traverse_inorder(print_node)


